### PR TITLE
Make `push_outlives_components` into a `TypeVisitor`

### PR DESCRIPTION
**Pull request by compiler-errors**
This involves removing the `visited: &mut SsoHashSet<GenericArg<'tcx>>` that is being passed around the `VerifyBoundCx`. The fact that we were using it when decomposing different type tests seems sketchy, so I don't think, though it may technically result in us registering more redundant outlives components 🤷 

I did end up deleting some of the comments that referred back to RFC 1214 during this refactor. I can add them back if you think they were useful.

r? lcnr

### CODE REVIEW

File: compiler/rustc_type_ir/src/outlives.rs [hunk 1]
```diff
@@ -56,215 +55,170 @@ pub enum Component<I: Interner> {
...
+                    // the rules OutlivesProjectionEnv,
+                    // OutlivesProjectionTraitDef, and
+                    // OutlivesProjectionComponents to regionck.
+                    self.out.push(Component::Alias(alias_ty));
+                } else {
+                    // fallback case: hard code
+                    // OutlivesProjectionComponents. Continue walking
+                    // through and constrain Pi.
+                    let mut subcomponents = smallvec![];
+                    compute_alias_components_recursive(self.tcx, ty, &mut subcomponents);
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/outlives.rs:153]: can you inline this function? It's very subtle that it uses a separate visitor/cache, so I would like to make this very explicit

File: compiler/rustc_type_ir/src/outlives.rs [hunk 2]
```diff
@@ -56,215 +55,170 @@ pub enum Component<I: Interner> {
...
-        ty::Bound(..) |
-        ty::Error(_) => {
-            // Trivial.
-        }
+            // Most types do not introduce any region binders, nor
+            // involve any other subtle cases, and so the WF relation
+            // simply constraints any regions referenced directly by
+            // the type and then visits the types that are lexically
+            // contained within. (The comments refer to relevant rules
+            // from RFC1214.)
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/outlives.rs:170]:
```suggestion
            // contained within.
```

File: compiler/rustc_type_ir/src/outlives.rs [hunk 3]
```diff
@@ -56,215 +55,170 @@ pub enum Component<I: Interner> {
...
+        match ct.kind() {
+            ty::ConstKind::Param(_)
+            | ty::ConstKind::Bound(_, _)
+            | ty::ConstKind::Infer(_)
+            | ty::ConstKind::Placeholder(_)
+            | ty::ConstKind::Error(_) => {
+                // Trivial
+            }
+            ty::ConstKind::Expr(..) | ty::ConstKind::Value(..) | ty::ConstKind::Unevaluated(..) => {
+                ct.super_visit_with(self);
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/outlives.rs:221]: 2 ideas: change visited to only cache types. Change visit const to just be `super_visit_with` 🤔 this is the meaning of this match after all

**Review by lcnr — COMMENTED** [review]
some nits and suggestions

**Comment by lcnr** [comment]
r=me after handling my comments in some way :3

File: compiler/rustc_type_ir/src/outlives.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/outlives.rs:153]: It's being called in two places -- here, and in `rustc_infer`. That would require us to expose the outlives visitor there to replicate this behavior, or just inline this one callsite? that seems weird imo

**Pull request merged**